### PR TITLE
Proposal: Use jsonpb to (un)marshall messages to/from JSON

### PR DESCRIPTION
**Pull request by pauln**
The stdlib encoding/json package isn't able to (un)marshal some advanced protobuf features such as `oneof` fields.  [golang/protobuf/jsonpb](http://github.com/golang/protobuf/jsonpb) should be used instead, as it follows the [canonical proto3 JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json) and therefore supports these features.

This change should allow `oneof` fields to be used, though the warning message added by #226 can probably be removed again (or degraded to debug level) as it (mostly) relates to the previous lack of support for `oneof` fields.

### CODE REVIEW

File: cmd/_integration-tests/transport/http_test.go
Comment (zaquestion) [inline cmd/_integration-tests/transport/http_test.go:611]: Can these be combined into 1 case statement? 
```
 case *pb.GetWithQueryResponse, *pb.GetWithRepeatedQueryResponse,...:
```

Comment (adamryman) [reply cmd/_integration-tests/transport/http_test.go:611]: You cannot:
```
./http_test.go:612: cannot use v (type interface {}) as type proto.Message in argument to jsonpb.UnmarshalString:
        interface {} does not implement proto.Message (missing ProtoMessage method)
```

By being only one type per case then the type of `v` is known. If you combine them into one then `v` is of type `interface{}`.

Comment (pauln) [reply cmd/_integration-tests/transport/http_test.go:611]: Yeah, I found that out the hard way - tried combining cases, still got an interface.  It's more verbose (and repetitive) than I'd like, but as `jsonpb.Unmarshal` wants a `proto.Message`, this was the shortest / simplest way I could find to do it.

**Comment by pauln** [comment]
I've just pushed another commit, as I discovered a regression wherein unknown keys in the request JSON resulted in an error from `jsonpb.UnmarshalString`, though they were previously silently ignored.  The workaround requires switching to a `jsonpb.Unmarshaler` so that we can set `AllowUnknownFields`.

File: gengokit/httptransport/templates/server.go
Comment (adamryman) [inline gengokit/httptransport/templates/server.go:177]: We are a little concerned about this. Could you point us at documentation that jsonpb, does *not* do html escaping? 

See here https://godoc.org/encoding/json#Encoder.SetEscapeHTML. We had a problem with this in production before.

We will be testing this in our staging environment for one of products here at Tune, as we do not have tests for the HTML escaping in json. I believe that I'll be able to do that before the end of the week. If you would like to add a test for HTML characters being escaped in json, it would be highly appreciated and give strong confidence for merging this.

Comment (pauln) [reply gengokit/httptransport/templates/server.go:177]: A quick test shows that it unfortunately _is_ doing html escaping.  I'll do a bit of digging to see if that's able to be changed.

Comment (pauln) [reply gengokit/httptransport/templates/server.go:177]: I can't see a way to disable the HTML escaping (`jsonpb` internally uses `encoding/json`, but doesn't currently have a way to call `SetEscapeHTML`), so I've filed an issue.

**Review by adamryman — DISMISSED** [review]
I'm putting a block on this PR till we resolve `encoder.SetEscapeHTML(false)` being removed and not having an equivalent in `jsonpb`.

[See this review in this PR](https://github.com/TuneLab/truss/pull/227/files#r132371930)
[See this issue in golang/protobuf](https://github.com/TuneLab/truss/pull/227/files#r132371930)

**Comment by pauln** [comment]
I've [submitted a PR](https://github.com/golang/protobuf/pull/409) which adds a `SkipEscapeHTML` option to `jsonpb.Marshaler` which should unblock this if/when merged.  It currently fails CI on Go1.6 due to lack of options on `json.Encoder`, so I'm not sure how soon they might be willing to merge it.

**Comment by zaquestion** [comment]
With truss now able to support `google.protobuf` types through gogo, these changes seem more relevant than ever!

For
```
message SearchResponse {
  int64 num_results = 1;
  repeated google.protobuf.Struct results = 2;

}
```

On master trying to marshal the `google.protobuf.Struct` type yields
```
$ curl 'localhost:5050/search' -d '{ "query": "foo" }'
{"num_results":1,"results":[{"fields":{"displayName":{"Kind":{"StringValue":"Denver, CO"}},"objectType":{"Kind":{"StringValue":"place"}}}}]}
```
regenerated with these changes merged in we get the clean response we want.
```
$ truss service.proto
$ curl 'localhost:5050/search' -d '{ "query": "foo" }'
{"numResults":"1","results":[{"displayName":"Denver, CO","objectType":"place"}]}
```
The response data is hardcoded and not actually based on the query or anything. I haven't tried in the other direction, but based on the code here and the above value add already, I think its worth pushing through. The original hold on escape html was more specific to TUNE and those services aren't in production any more, I'm on board with starting fresh.

**Comment by lelandbatey** [comment]
The original HTML escape case that we where hung up on is something we don't really care about any longer. The improvements from `jsonpb` vastly outweigh the edge cases, so I'm very in favor of merging this PR.

**Review by zaquestion — APPROVED** [review]
Approved in conjunction with @lelandbatey

**Comment by zaquestion** [comment]
@pauln After all this time, we've come to our senses! Excited to have this in :D

**Pull request merged**
